### PR TITLE
DEVDOC: Add life of a gfxtrace

### PR DESCRIPTION
**Pull request by hevrard**
Give an overview of how AGI captures and replays graphics API
commands. This can help developers build a mental model to navigate
the code.

Bug: n/a

### CODE REVIEW

File: DEVDOC.md [hunk 1]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+
+This is an overview of how AGI captures graphics API calls, stores them into a
+gfxtrace file, and process that file to replay the calls.
+
+This is oriented to AGI developers who need to work with AGI internals. It is
+meant to help build a high-level mental model that facilitates navigating the
+actual source code. It references concepts and code modules that are unlikely to
+change anytime soon, and it tries not to reference details that are likely to
+change. You are encouraged to check the source code while reading this, to
+clarify how the various parts actually fits together. If you spot any
```
Comment (johnplate) [inline DEVDOC.md:412]: I would use "fit" here, because parts is plural.

File: DEVDOC.md [hunk 2]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+discrepancy, please update this doc, but please keep it high-level.
+
+AGI is architectured to be API-agnostic in order to support multiple graphics
+API. GAPID, from which AGI was forked, supports GLES and Vulkan. AGI main focus
```
Comment (johnplate) [inline DEVDOC.md:416]: APIs should be plural. And I think better use "AGI's main focus", but better double check with a native English speaker.

File: DEVDOC.md [hunk 3]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+and then a new GAPIC that connects to this GAPIS. GAPIS may then itself start
+GAPII or GAPIR instances.
+
+To capture and replay on Android, AGI has APKs (one per ABI, e.g.
+`gapid-arm64-v8a.apk`) that embeds GAPII and GAPIR.
+
+### Create a gfxtrace: capture, serialize and store Vulkan calls
+
+AGI uses a Vulkan layer to intercept and capture the Vulkan calls emitted by an
+application, and store them into a gfxtrace file. Be aware that in the code
```
Comment (johnplate) [inline DEVDOC.md:455]: stores

File: DEVDOC.md [hunk 4]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+base, the act of intercepting graphics API calls may be referred by the words
+"capture", "trace", "intercept" and "spy".
+
+For a Vulkan capture on Android, the main steps are:
+
+1. GAPIS does a few adb commands to edit global settings that tell the Android
```
Comment (johnplate) [inline DEVDOC.md:461]: "does commands" sounds weird, better "issues commands".

File: DEVDOC.md [hunk 5]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+   Vulkan loader to inserts AGI's Vulkan layer when starting the app to capture.
```
Comment (johnplate) [inline DEVDOC.md:462]: to insert

File: DEVDOC.md [hunk 6]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+   This layer is called `GraphicsSpy`, it is implemented by
```
Comment (johnplate) [inline DEVDOC.md:463]: This layer is called `GraphicsSpy` and is implemented by

File: DEVDOC.md [hunk 7]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+   `libVkLayer_GraphicsSpy.so` which is mostly a wrapper around `libgapii.so`.
+
+2. The capture layer and GAPIS establish a TCP connection over ADB. The
+   `gapii::Spy::Spy()` creator contains the logic to establish this connection.
+
+3. The capture layer monitors every Vulkan API calls, and shadows the Vulkan
```
Comment (johnplate) [inline DEVDOC.md:469]: every Vulkan API call

File: DEVDOC.md [hunk 8]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+   struct as arguments: the content of these structures is read by the driver.
+
+3. The actual call to the driver, and its return value (if any).
+
+4. Zero or more memory observations that represent memory that the driver may
+   have written to during its processing of the command. For instance, a
+   `vkCreateBuffer` call writes the handle of the newly create buffer to the
+   memory pointed at by its `VkBuffer* pBuffer` argument.
+
+Note that on multihreaded apps, the object groups may be interleaved.
```
Comment (johnplate) [inline DEVDOC.md:553]: Missing 't' in multithreaded

File: DEVDOC.md [hunk 9]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+change. You are encouraged to check the source code while reading this, to
+clarify how the various parts actually fits together. If you spot any
+discrepancy, please update this doc, but please keep it high-level.
+
+AGI is architectured to be API-agnostic in order to support multiple graphics
+API. GAPID, from which AGI was forked, supports GLES and Vulkan. AGI main focus
+is Vulkan on Android, so all examples in this doc are related to Vulkan on
+Android.
+
+### AGI overall architecture
```
Comment (johnplate) [inline DEVDOC.md:420]: Maybe better to use 'AGI's overall architecture'

File: DEVDOC.md [hunk 10]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+
+- GAPIS (Graphics API Server, Go code under `gapis/`, see `gapis/README.md`) is
+  the main component, running on the developer desktop/laptop. Any complex logic
+  is meant to be implemented in GAPIS, while the other components are meant to
+  be kept as simple as possible. The protobuf interface of GAPIS is defined in
+  `gapis/service/service.proto`.
+
+- GAPII (Graphics API Interceptor, mostly C++ code under `gapii/`, see
+  `gapii/README.md`) is the component responsible for intercepting graphics API
+  calls during capture.  It interacts with GAPIS via a dedicated protocol.
```
Comment (johnplate) [inline DEVDOC.md:432]: There are many double spaces in the document, like here between capture and It. These have a special meaning if they are at the end of a line, so they should be avoided if they are not intentional. Please search them all and replace with single spaces.

File: DEVDOC.md [hunk 11]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+### GAPIR executes the replay instructions
+
+GAPIS uses the GAPIR protobuf service (`gapir/replay_service/service.proto`) to
+request the replay of a payload. This payload contains replay opcodes, and
+references to the required replay resources (i.e. the raw bytes of e.g. buffer
+contents).
+
+The resources are not directly embedded in the replay payload: GAPIR lazily
+requests them to GAPIS during the replay. On Android, GAPIR runs on the device
+and communicate with GAPIS over ADB. Pushing the resources from GAPIS to GAPIR
```
Comment (johnplate) [inline DEVDOC.md:634]: communicates

File: DEVDOC.md [hunk 12]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+to be replayed even before the user asks for a replay.
+
+Depending on the replay request, the capture commands may be transformed to
+produce a relevant replay payload. Most of the time, these transformations only
+need to be applied on the capture commands, not on the initial commands. Hence,
+when waiting for user input, it makes sense to replay the default initial
+commands to rebuilt the initial state: once the user requests a specific replay,
+only the (transformed) capture commands needs to be replayed.
+
+If the user-requested replay do require to transform the initial commands, then
```
Comment (johnplate) [inline DEVDOC.md:658]: does

File: DEVDOC.md [hunk 13]
```diff
@@ -396,3 +396,265 @@ bazel-bin/gapii/cc/vulkan_spy_1.cpp
...
+the pre-warm replay is abandonned and a new replay of the transformed inital
+commands and then the transformed capture commands is executed.
```
Comment (johnplate) [inline DEVDOC.md:660]: are executed

**Review by johnplate — CHANGES_REQUESTED** [review]
Great work! Just a few typos and similar.

File: DEVDOC.md (hunk 1 above)
Comment (hevrard) [reply DEVDOC.md:412]: done

File: DEVDOC.md (hunk 3 above)
Comment (hevrard) [reply DEVDOC.md:455]: done

File: DEVDOC.md (hunk 2 above)
Comment (hevrard) [reply DEVDOC.md:416]: done

File: DEVDOC.md (hunk 4 above)
Comment (hevrard) [reply DEVDOC.md:461]: done

File: DEVDOC.md (hunk 5 above)
Comment (hevrard) [reply DEVDOC.md:462]: done

File: DEVDOC.md (hunk 6 above)
Comment (hevrard) [reply DEVDOC.md:463]: done

File: DEVDOC.md (hunk 7 above)
Comment (hevrard) [reply DEVDOC.md:469]: done

File: DEVDOC.md (hunk 8 above)
Comment (hevrard) [reply DEVDOC.md:553]: done

File: DEVDOC.md (hunk 9 above)
Comment (hevrard) [reply DEVDOC.md:420]: done

File: DEVDOC.md (hunk 10 above)
Comment (hevrard) [reply DEVDOC.md:432]: done

File: DEVDOC.md (hunk 11 above)
Comment (hevrard) [reply DEVDOC.md:634]: done

**Review by johnplate — APPROVED** [review]

**Pull request merged**